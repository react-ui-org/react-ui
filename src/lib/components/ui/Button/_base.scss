@use '../../../styles/tools/accessibility';
@use '../../../styles/tools/breakpoint';
@use 'settings';
@use 'theme';
@use 'tools';

.root {
  @include tools.button();
}

.label {
  display: block;
}

.beforeLabel,
.afterLabel,
.startCorner,
.endCorner,
.loadingIcon {
  display: flex;
  align-items: baseline;
  justify-content: center;
}

.beforeLabel {
  margin-right: settings.$icon-spacing;
}

.afterLabel,
.loadingIcon {
  margin-left: settings.$icon-spacing;
}

.startCorner,
.endCorner {
  position: absolute;
  top: -0.35rem;
}

.startCorner {
  left: 0;
  margin-left: -0.35rem;
}

.endCorner {
  right: 0;
  margin-right: -0.35rem;
}

.rootSizeSmall {
  @include tools.button-size(small);
}

.rootSizeMedium {
  @include tools.button-size(medium);
}

.rootSizeLarge {
  @include tools.button-size(large);
}

.rootBlock {
  width: 100%;
}

.rootBlock.withLabelHiddenMobile {
  @include breakpoint.up(settings.$breakpoint) {
    width: 100%;
  }
}

.rootGrouped {
  z-index: map-get(settings.$group-z-indexes, button);

  &:not(:first-child) {
    margin-left: var(--local-gap);
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }

  &:not(:last-child) {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }

  &:focus,
  &:not(:disabled):hover {
    z-index: map-get(settings.$group-z-indexes, button-hover);
  }
}

.rootGrouped .startCorner,
.rootGrouped .endCorner {
  z-index: map-get(settings.$group-z-indexes, button-overflowing-elements);
}

.withLabelHidden .label,
.withLabelHiddenMobile .label {
  @include accessibility.hide-text();
}

.withLabelHiddenMobile .label {
  @include breakpoint.up(settings.$breakpoint) {
    position: relative;
    width: auto;
    height: auto;
  }
}

.withLabelHidden .beforeLabel,
.withLabelHidden .afterLabel,
.withLabelHiddenMobile .beforeLabel,
.withLabelHiddenMobile .afterLabel {
  margin-right: 0;
  margin-left: 0;
}

.withLabelHiddenMobile .beforeLabel {
  @include breakpoint.up(settings.$breakpoint) {
    margin-right: settings.$icon-spacing;
  }
}

.withLabelHiddenMobile .afterLabel {
  @include breakpoint.up(settings.$breakpoint) {
    margin-left: settings.$icon-spacing;
  }
}

.withLabelHidden.isRootLoading .loadingIcon,
.withLabelHiddenMobile.isRootLoading .loadingIcon {
  margin-left: 0;
}

.withLabelHidden.isRootLoading .beforeLabel,
.withLabelHidden.isRootLoading .afterLabel,
.withLabelHiddenMobile.isRootLoading .beforeLabel,
.withLabelHiddenMobile.isRootLoading .afterLabel {
  display: none;
}

.withLabelHiddenMobile.isRootLoading .beforeLabel,
.withLabelHiddenMobile.isRootLoading .afterLabel {
  @include breakpoint.up(settings.$breakpoint) {
    display: flex;
  }
}

.withLabelHiddenMobile.isRootLoading .loadingIcon {
  @include breakpoint.up(settings.$breakpoint) {
    margin-left: settings.$icon-spacing;
  }
}
