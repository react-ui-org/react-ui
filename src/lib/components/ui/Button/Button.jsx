import PropTypes from 'prop-types';
import React from 'react';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './Button.scss';

export const Button = (props) => {
  const {
    afterLabel,
    beforeLabel,
    block,
    clickHandler,
    disabled,
    endCorner,
    forwardedRef,
    grouped,
    id,
    label,
    labelVisibility,
    loadingIcon,
    priority,
    size,
    startCorner,
    type,
    variant,
  } = props;

  let priorityClass = '';
  let sizeClass = '';
  let variantClass = '';
  let blockClass = '';
  let groupedClass = '';
  let labelVisibilityClass = '';
  let loadingClass = '';

  if (priority === 'default') {
    priorityClass = styles.priorityDefault;
  } else if (priority === 'outline') {
    priorityClass = styles.priorityOutline;
  } else if (priority === 'flat') {
    priorityClass = styles.priorityFlat;
  } else if (priority === 'link') {
    priorityClass = styles.priorityLink;
  }

  if (priority !== 'link') {
    if (size === 'small') {
      sizeClass = styles.sizeSmall;
    } else if (size === 'medium') {
      sizeClass = styles.sizeMedium;
    } else if (size === 'large') {
      sizeClass = styles.sizeLarge;
    }

    if (variant === 'primary') {
      variantClass = styles.variantPrimary;
    } else if (variant === 'secondary') {
      variantClass = styles.variantSecondary;
    } else if (variant === 'success') {
      variantClass = styles.variantSuccess;
    } else if (variant === 'warning') {
      variantClass = styles.variantWarning;
    } else if (variant === 'danger') {
      variantClass = styles.variantDanger;
    } else if (variant === 'dark') {
      variantClass = styles.variantDark;
    }

    if (block) {
      blockClass = styles.isRootBlock;
    }

    if (grouped) {
      groupedClass = styles.isRootGrouped;
    }

    if (labelVisibility === 'desktop') {
      labelVisibilityClass = styles.withLabelHiddenMobile;
    } else if (labelVisibility === 'none') {
      labelVisibilityClass = styles.withLabelHidden;
    }

    if (loadingIcon) {
      loadingClass = styles.isRootLoading;
    }
  }

  const propsToTransfer = transferProps(
    props,
    ['afterLabel', 'beforeLabel', 'block', 'clickHandler', 'disabled', 'endCorner', 'grouped', 'id',
      'label', 'labelVisibility', 'loadingIcon', 'priority', 'size', 'startCorner', 'type', 'variant'],
  );

  return (
    <button
      {...propsToTransfer}
      className={(`
        ${styles.root}
        ${priorityClass}
        ${sizeClass}
        ${variantClass}
        ${blockClass}
        ${groupedClass}
        ${labelVisibilityClass}
        ${loadingClass}
      `).trim()}
      disabled={disabled || !!loadingIcon}
      id={id}
      onClick={clickHandler}
      ref={forwardedRef}
      type={type}
    >
      {priority !== 'link' && startCorner && (
        <span className={styles.startCorner}>
          {startCorner}
        </span>
      )}
      {beforeLabel && (
        <span className={styles.beforeLabel}>
          {beforeLabel}
        </span>
      )}
      <span
        className={styles.label}
        {...(id && { id: `${id}__labelText` })}
      >
        {label}
      </span>
      {afterLabel && (
        <span className={styles.afterLabel}>
          {afterLabel}
        </span>
      )}
      {priority !== 'link' && endCorner && (
        <span className={styles.endCorner}>
          {endCorner}
        </span>
      )}
      {priority !== 'link' && loadingIcon && (
        <span className={styles.loadingIcon}>
          {loadingIcon}
        </span>
      )}
    </button>
  );
};

Button.defaultProps = {
  afterLabel: null,
  beforeLabel: null,
  block: false,
  clickHandler: null,
  disabled: false,
  endCorner: null,
  forwardedRef: undefined,
  grouped: false,
  id: undefined,
  labelVisibility: 'all',
  loadingIcon: null,
  priority: 'default',
  size: 'medium',
  startCorner: null,
  type: 'button',
  variant: 'primary',
};

Button.propTypes = {
  afterLabel: PropTypes.node,
  beforeLabel: PropTypes.node,
  block: PropTypes.bool,
  clickHandler: PropTypes.func,
  disabled: PropTypes.bool,
  endCorner: PropTypes.node,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  grouped: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  labelVisibility: PropTypes.oneOf(['all', 'desktop', 'none']),
  loadingIcon: PropTypes.node,
  priority: PropTypes.oneOf(['default', 'outline', 'flat', 'link']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  startCorner: PropTypes.node,
  type: PropTypes.oneOf(['button', 'submit']),
  variant: PropTypes.oneOf(['primary', 'secondary', 'success', 'warning', 'danger', 'dark']),
};

export default withForwardedRef(Button);
