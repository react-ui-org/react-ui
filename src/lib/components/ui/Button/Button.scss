@import '../../../styles/settings/colors';
@import '../../../styles/tools/accessibility';
@import '../../../styles/tools/breakpoints';
@import '../../../styles/tools/forms/relationships';
@import 'mixins';
@import 'variables';

.root {
  @include button();
  @include form-field-horizontal-neighbor();
}

.label {
  display: block;
}

.beforeLabel,
.afterLabel,
.startCorner,
.endCorner,
.loadingIcon {
  display: flex;
  align-items: baseline;
  justify-content: center;
}

.beforeLabel {
  margin-right: $button-icon-offset;
}

.afterLabel,
.loadingIcon {
  margin-left: $button-icon-offset;
}

.startCorner,
.endCorner {
  position: absolute;
  top: -0.35rem;
}

.startCorner {
  left: 0;
  margin-left: -0.35rem;
}

.endCorner {
  right: 0;
  margin-right: -0.35rem;
}

.sizeSmall {
  @include button-size(small);
}

.sizeMedium {
  @include button-size(medium);
}

.sizeLarge {
  @include button-size(large);
}

.isRootBlock {
  @include form-field-vertical-neighbor();

  width: 100%;
}

.isRootBlock.withLabelHiddenMobile {
  @include breakpoint-up($button-breakpoint) {
    width: 100%;
  }
}

.isRootGrouped {
  &:not(:first-child) {
    margin-left: calc(-1 * #{$button-border-width});
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }

  &:not(:last-child) {
    margin-right: 0;
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
}

.withLabelHidden .label,
.withLabelHiddenMobile .label {
  @include accessibility-hide-text();
}

.withLabelHiddenMobile .label {
  @include breakpoint-up($button-breakpoint) {
    position: relative;
    width: auto;
    height: auto;
  }
}

.withLabelHidden .beforeLabel,
.withLabelHidden .afterLabel,
.withLabelHiddenMobile .beforeLabel,
.withLabelHiddenMobile .afterLabel {
  margin-right: 0;
  margin-left: 0;
}

.withLabelHiddenMobile .beforeLabel {
  @include breakpoint-up($button-breakpoint) {
    margin-right: $button-icon-offset;
  }
}

.withLabelHiddenMobile .afterLabel {
  @include breakpoint-up($button-breakpoint) {
    margin-left: $button-icon-offset;
  }
}

// Default priority variants
.priorityDefault.variantPrimary {
  @include button-variant(default, primary);
}

.priorityDefault.variantSecondary {
  @include button-variant(default, secondary);
}

.priorityDefault.variantSuccess {
  @include button-variant(default, success);
}

.priorityDefault.variantWarning {
  @include button-variant(default, warning);
}

.priorityDefault.variantDanger {
  @include button-variant(default, danger);
}

.priorityDefault.variantDark {
  @include button-variant(default, dark);
}

// Outline priority variants
.priorityOutline.variantPrimary {
  @include button-variant(outline, primary);
}

.priorityOutline.variantSecondary {
  @include button-variant(outline, secondary);
}

.priorityOutline.variantSuccess {
  @include button-variant(outline, success);
}

.priorityOutline.variantWarning {
  @include button-variant(outline, warning);
}

.priorityOutline.variantDanger {
  @include button-variant(outline, danger);
}

.priorityOutline.variantDark {
  @include button-variant(outline, dark);
}

// Flat priority variants
.priorityFlat.variantPrimary {
  @include button-variant(flat, primary);
}

.priorityFlat.variantSecondary {
  @include button-variant(flat, secondary);
}

.priorityFlat.variantSuccess {
  @include button-variant(flat, success);
}

.priorityFlat.variantWarning {
  @include button-variant(flat, warning);
}

.priorityFlat.variantDanger {
  @include button-variant(flat, danger);
}

.priorityFlat.variantDark {
  @include button-variant(flat, dark);
}

// Grouped buttons
.priorityDefault.isRootGrouped:not(:first-child)::before,
.priorityFlat.isRootGrouped:not(:first-child)::before {
  content: '';
  position: absolute;
  top: calc(-1 * #{$button-border-width});
  bottom: calc(-1 * #{$button-border-width});
  left: calc(-1 * #{$button-border-width});
  border-left: $button-border-width solid $color-white;
}

.priorityFlat.isRootGrouped:not(:first-child)::before {
  border-left-color: currentColor;
}
