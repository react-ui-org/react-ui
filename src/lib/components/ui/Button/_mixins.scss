// 1. Get ready for position of corner elements and group separator.
// 2. Icon buttons should be squares.

@import '../../../styles/tools/explode';
@import '../../../styles/tools/forms/states';
@import '../../../styles/tools/transitions';
@import 'variables';
@import 'theme';

@mixin button() {
  @include transition((opacity, color, border, background, box-shadow));

  position: relative; // 1.
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-weight: $button-font-weight;
  line-height: $button-line-height;
  font-family: $button-font-family;
  text-align: center;
  text-overflow: ellipsis;
  text-decoration: none;
  text-transform: $button-text-transform;
  white-space: nowrap;
  vertical-align: middle;
  border: $button-border-width solid transparent;
  border-radius: $button-border-radius;
  user-select: none;
  cursor: pointer;

  &:hover,
  &:focus {
    text-decoration: none;
  }

  &:disabled {
    @include form-field-disabled();
  }

  &:-moz-focusring {
    outline: 0;
  }
}

@mixin button-size($size) {
  $properties: map-get($button-sizes, $size);

  height: map-get($properties, height);
  padding: 0 map-get($properties, padding-x);
  font-size: map-get($properties, font-size);

  &.withLabelHidden,
  &.withLabelHiddenMobile {
    width: map-get($properties, height); // 2.
    padding-right: 0;
    padding-left: 0;
  }

  &.withLabelHiddenMobile {
    @include breakpoint-up($button-breakpoint) {
      width: auto;
      padding-right: map-get($properties, padding-x);
      padding-left: map-get($properties, padding-x);
    }
  }
}

@mixin button-variant($priority, $variant) {
  $priority-variants: map-get($button-variants, $priority);
  $variant-states: map-get($priority-variants, $variant);

  @include explode(map-get($variant-states, default));

  &:hover,
  &:focus {
    @include explode(map-get($variant-states, hover));
  }

  &:active {
    @include explode(map-get($variant-states, active));
  }

  &:disabled {
    @include explode(map-get($variant-states, default));
  }
}
