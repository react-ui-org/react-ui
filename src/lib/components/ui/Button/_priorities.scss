@use 'settings';
@use 'theme';
@use 'tools';

.rootPriorityFilled.rootColorPrimary {
  @include tools.button-color(filled, primary);
}

.rootPriorityFilled.rootColorSecondary {
  @include tools.button-color(filled, secondary);
}

.rootPriorityFilled.rootColorSuccess {
  @include tools.button-color(filled, success);
}

.rootPriorityFilled.rootColorWarning {
  @include tools.button-color(filled, warning);
}

.rootPriorityFilled.rootColorDanger {
  @include tools.button-color(filled, danger);
}

.rootPriorityFilled.rootColorHelp {
  @include tools.button-color(filled, help);
}

.rootPriorityFilled.rootColorInfo {
  @include tools.button-color(filled, info);
}

.rootPriorityFilled.rootColorNote {
  @include tools.button-color(filled, note);
}

.rootPriorityFilled.rootColorLight {
  @include tools.button-color(filled, light);
}

.rootPriorityFilled.rootColorDark {
  @include tools.button-color(filled, dark);
}

.rootPriorityOutline.rootColorPrimary {
  @include tools.button-color(outline, primary);
}

.rootPriorityOutline.rootColorSecondary {
  @include tools.button-color(outline, secondary);
}

.rootPriorityOutline.rootColorSuccess {
  @include tools.button-color(outline, success);
}

.rootPriorityOutline.rootColorWarning {
  @include tools.button-color(outline, warning);
}

.rootPriorityOutline.rootColorDanger {
  @include tools.button-color(outline, danger);
}

.rootPriorityOutline.rootColorHelp {
  @include tools.button-color(outline, help);
}

.rootPriorityOutline.rootColorInfo {
  @include tools.button-color(outline, info);
}

.rootPriorityOutline.rootColorNote {
  @include tools.button-color(outline, note);
}

.rootPriorityOutline.rootColorLight {
  @include tools.button-color(outline, light);
}

.rootPriorityOutline.rootColorDark {
  @include tools.button-color(outline, dark);
}

.rootPriorityFlat.rootColorPrimary {
  @include tools.button-color(flat, primary);
}

.rootPriorityFlat.rootColorSecondary {
  @include tools.button-color(flat, secondary);
}

.rootPriorityFlat.rootColorSuccess {
  @include tools.button-color(flat, success);
}

.rootPriorityFlat.rootColorWarning {
  @include tools.button-color(flat, warning);
}

.rootPriorityFlat.rootColorDanger {
  @include tools.button-color(flat, danger);
}

.rootPriorityFlat.rootColorHelp {
  @include tools.button-color(flat, help);
}

.rootPriorityFlat.rootColorInfo {
  @include tools.button-color(flat, info);
}

.rootPriorityFlat.rootColorNote {
  @include tools.button-color(flat, note);
}

.rootPriorityFlat.rootColorLight {
  @include tools.button-color(flat, light);
}

.rootPriorityFlat.rootColorDark {
  @include tools.button-color(flat, dark);
}

.rootPriorityLink {
  @include tools.button-link();
}

.rootPriorityFilled.rootGrouped:not(:first-child)::before,
.rootPriorityFlat.rootGrouped:not(:first-child)::before {
  content: '';
  position: absolute;
  top: calc(-1 * #{theme.$border-width});
  bottom: calc(-1 * #{theme.$border-width});
  left: calc(-1 * #{theme.$border-width});
  z-index: map-get(settings.$group-z-indexes, separator);
  border-left: var(--local-separator-width) solid var(--local-separator-color);
  transform: translateX(calc(-0.5 * var(--local-gap) - 50%));
}

.rootPriorityFilled.rootGrouped:not(:first-child) {
  --local-gap: #{theme.$group-filled-gap};
  --local-separator-width: #{theme.$group-filled-separator-width};
  --local-separator-color: #{theme.$group-filled-separator-color};
}

.rootPriorityFlat.rootGrouped:not(:first-child) {
  --local-gap: #{theme.$group-flat-gap};
  --local-separator-width: #{theme.$group-flat-separator-width};
  --local-separator-color: #{theme.$group-flat-separator-color};
}

.rootPriorityOutline.rootGrouped:not(:first-child) {
  --local-gap: #{theme.$group-outline-gap};
}
