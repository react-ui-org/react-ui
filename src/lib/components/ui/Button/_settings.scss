@use '../../../styles/settings/typography';
@use '../../../styles/tools/spacing';

$font-family: typography.$font-family-base;
$line-height: typography.$line-height-base;
$icon-spacing: spacing.of(2);
$breakpoint: lg;

$themeable-variants: (primary, secondary, success, warning, danger, dark);
$themeable-states: (default, hover, active, disabled);
$themeable-priority-properties: (
  filled: (color, border-color, background, box-shadow),
  outline: (color, border-color, background),
  flat: (color, background),
);
