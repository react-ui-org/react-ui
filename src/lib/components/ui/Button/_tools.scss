// 1. Due to complexity, it's easier and safer to resolve variant custom properties dynamically with
//    a mixin. This way it's also easier to make theming adjustments to all supported modifications
//    in one place. This applies to the following button priorities:
//
//    - filled
//    - outline
//    - flat
//
//    The “link” priority is treated separately as it doesn't share much of appearance with other
//    variants listed above.
//
// 2. Disabled state theming is optional so `default` theme options are used if no `disabled`
//    styling is provided for specific variant.
//
// 3. Get ready for position of corner elements and group separator.
//
// 4. Icon buttons should appear as squares, that's why width and height is equal here.
//
// 5. Unset undesired properties so the link button appears as link while it keeps some button
//    settings. It might be tempting to use something like `all: initial` or `all: inherit` but it
//    actually causes more harm rather than a useful reset.

@use 'sass:math';
@use 'sass:list';
@use '../../../styles/tools/breakpoint';
@use '../../../styles/tools/transition';
@use 'settings';
@use 'theme';

// 1.
@mixin _get-themeable-properties($priority, $variant, $state) {
  @if (not list.index(map-keys(settings.$themeable-priority-properties), $priority)) {
    @error "Invalid priority specified! #{$priority} doesn't exist. "
      + "Choose one of #{settings.$themeable-priority-properties}.";
  }

  @if (not list.index(settings.$themeable-variants, $variant)) {
    @error "Invalid variant specified! #{$variant} doesn't exist. "
      + "Choose one of #{settings.$themeable-variants}.";
  }

  @if (not list.index(settings.$themeable-states, $state)) {
    @error "Invalid state specified! #{$state} doesn't exist. "
      + "Choose one of #{settings.$themeable-states}.";
  }

  $properties: map-get(settings.$themeable-priority-properties, $priority);

  // 2.
  @if ($state == 'disabled') {
    @each $property in $properties {
      --rui-local-#{$property}:
        var(
          --rui-Button--#{$priority}--#{$variant}--#{$state}__#{$property},
          var(--rui-Button--#{$priority}--#{$variant}--default__#{$property})
        );
    }
  }

  @else {
    @each $property in $properties {
      --rui-local-#{$property}: var(--rui-Button--#{$priority}--#{$variant}--#{$state}__#{$property});
    }
  }
}

@mixin _unset-button-appearance() {
  width: unset;
  height: unset;
  padding: unset;
  font-weight: unset;
  line-height: unset;
  font-family: unset;
  text-transform: unset;
  vertical-align: unset;
  border: unset;
  border-radius: unset;
  box-shadow: unset;
  user-select: unset;
  transition: unset;
}

@mixin button() {
  @include transition.add((opacity, color, border-color, background-color, box-shadow));

  position: relative; // 3.
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: var(--rui-local-width, auto);
  height: var(--rui-local-height, auto);
  padding: var(--rui-local-padding);
  font: theme.$font-weight list.slash(var(--rui-local-font-size), settings.$line-height) settings.$font-family;
  letter-spacing: theme.$letter-spacing;
  text-align: center;
  text-overflow: ellipsis;
  text-decoration: none;
  text-transform: theme.$text-transform;
  vertical-align: middle;
  color: var(--rui-local-color);
  border: theme.$border-width solid var(--rui-local-border-color, transparent);
  border-radius: theme.$border-radius;
  background: var(--rui-local-background, transparent);
  box-shadow: var(--rui-local-box-shadow, none);
  user-select: none;
  cursor: pointer;

  &:hover,
  &:focus {
    text-decoration: none;
  }

  &:disabled {
    opacity: theme.$disabled-opacity;
    cursor: theme.$disabled-cursor;
  }

  &:-moz-focusring {
    outline: 0;
  }
}

@mixin button-size($size) {
  $properties: map-get(theme.$sizes, $size);

  --rui-local-height: #{map-get($properties, height)};
  --rui-local-padding: 0 #{map-get($properties, padding-x)};
  --rui-local-font-size: #{map-get($properties, font-size)};

  &.withLabelHidden,
  &.withLabelHiddenMobile {
    --rui-local-padding: 0;
    --rui-local-width: #{map-get($properties, height)}; // 4.
  }

  &.withLabelHiddenMobile {
    @include breakpoint.up(settings.$breakpoint) {
      --rui-local-padding: 0 #{map-get($properties, padding-x)};
      --rui-local-width: auto;
    }
  }
}

@mixin button-variant($priority, $variant) {
  @include _get-themeable-properties($priority, $variant, default);

  &:disabled {
    @include _get-themeable-properties($priority, $variant, disabled);
  }

  &:not(:disabled):hover,
  &:not(:disabled):focus {
    @include _get-themeable-properties($priority, $variant, hover);
  }

  &:not(:disabled):active {
    @include _get-themeable-properties($priority, $variant, active);
  }
}

@mixin button-link() {
  @include _unset-button-appearance(); // 5.

  text-decoration: map-get(theme.$link-priority-properties, default, text-decoration);
  color: map-get(theme.$link-priority-properties, default, color);

  &:not(:disabled):hover,
  &:not(:disabled):focus {
    text-decoration: map-get(theme.$link-priority-properties, hover, text-decoration);
    color: map-get(theme.$link-priority-properties, hover, color);
  }
}
