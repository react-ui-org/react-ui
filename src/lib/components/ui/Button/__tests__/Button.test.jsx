import React from 'react';
import {
  render,
  screen,
  within,
} from '@testing-library/react';
import sinon from 'sinon';
import userEvent from '@testing-library/user-event';
import { colorPropTest } from '../../../../../../tests/propTests/colorPropTest';
import { forwardedRefPropTest } from '../../../../../../tests/propTests/forwardedRefPropTest';
import { labelPropTest } from '../../../../../../tests/propTests/labelPropTest';
import { sizePropTest } from '../../../../../../tests/propTests/sizePropTest';
import { ButtonGroupContext } from '../../ButtonGroup';
import { Button } from '../Button';

const mandatoryProps = {
  label: 'label',
};

describe('rendering', () => {
  it.each([
    [
      { block: true },
      (rootElement) => expect(rootElement).toHaveClass('rootBlock'),
    ],
    [
      { block: false },
      (rootElement) => expect(rootElement).not.toHaveClass('rootBlock'),
    ],
    [
      { disabled: true },
      (rootElement) => expect(rootElement).toBeDisabled(),
    ],
    [
      { disabled: false },
      (rootElement) => expect(rootElement).not.toBeDisabled(),
    ],
    [
      { priority: 'filled' },
      (rootElement) => expect(rootElement).toHaveClass('rootPriorityFilled'),
    ],
    [
      { priority: 'outline' },
      (rootElement) => expect(rootElement).toHaveClass('rootPriorityOutline'),
    ],
    [
      { priority: 'flat' },
      (rootElement) => expect(rootElement).toHaveClass('rootPriorityFlat'),
    ],
    ...sizePropTest,
  ])('renders with ButtonGroup props: "%s"', (testedProps, assert) => {
    const dom = render((
      <ButtonGroupContext.Provider
        value={{ ...testedProps }}
      >
        <Button
          {...mandatoryProps}
        />
      </ButtonGroupContext.Provider>
    ));

    assert(dom.container.firstChild);
  });

  it.each([
    [
      { afterLabel: <div>after label</div> },
      (rootElement) => expect(within(rootElement).getByText('after label')),
    ],
    [
      { beforeLabel: <div>before label</div> },
      (rootElement) => expect(within(rootElement).getByText('before label')),
    ],
    [
      { block: true },
      (rootElement) => expect(rootElement).toHaveClass('rootBlock'),
    ],
    [
      { block: false },
      (rootElement) => expect(rootElement).not.toHaveClass('rootBlock'),
    ],
    ...colorPropTest,
    [
      { disabled: true },
      (rootElement) => expect(rootElement).toBeDisabled(),
    ],
    [
      { disabled: false },
      (rootElement) => expect(rootElement).not.toBeDisabled(),
    ],
    [
      { endCorner: <div>corner text</div> },
      (rootElement) => expect(within(rootElement).getByText('corner text')),
    ],
    ...forwardedRefPropTest(React.createRef()),
    [
      { id: 'id' },
      (rootElement) => {
        expect(rootElement).toHaveAttribute('id', 'id');
        expect(within(rootElement).getByText('label')).toHaveAttribute('id', 'id__labelText');
      },
    ],
    ...labelPropTest,
    [
      { labelVisibility: 'sm' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelVisibleSm'),
    ],
    [
      { labelVisibility: 'md' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelVisibleMd'),
    ],
    [
      { labelVisibility: 'lg' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelVisibleLg'),
    ],
    [
      { labelVisibility: 'xl' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelVisibleXl'),
    ],
    [
      { labelVisibility: 'xxl' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelVisibleXxl'),
    ],
    [
      { labelVisibility: 'xxxl' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelVisibleXxxl'),
    ],
    [
      { labelVisibility: 'none' },
      (rootElement) => expect(rootElement).toHaveClass('withLabelHidden'),
    ],
    [
      { feedbackIcon: <div>feedback icon</div> },
      (rootElement) => {
        expect(within(rootElement).getByText('feedback icon'));
        expect(rootElement).toBeDisabled();
        expect(rootElement).toHaveClass('hasRootFeedback');
      },
    ],
    [
      { priority: 'filled' },
      (rootElement) => expect(rootElement).toHaveClass('rootPriorityFilled'),
    ],
    [
      { priority: 'outline' },
      (rootElement) => expect(rootElement).toHaveClass('rootPriorityOutline'),
    ],
    [
      { priority: 'flat' },
      (rootElement) => expect(rootElement).toHaveClass('rootPriorityFlat'),
    ],
    ...sizePropTest,
    [
      { startCorner: <div>corner text</div> },
      (rootElement) => expect(within(rootElement).getByText('corner text')),
    ],
    [
      { type: 'button' },
      (rootElement) => expect(rootElement).toHaveAttribute('type', 'button'),
    ],
    [
      { type: 'submit' },
      (rootElement) => expect(rootElement).toHaveAttribute('type', 'submit'),
    ],
  ])('renders with props: "%s"', (testedProps, assert) => {
    const dom = render((
      <Button
        {...mandatoryProps}
        {...testedProps}
      />
    ));

    assert(dom.container.firstChild);
  });
});

describe('functionality', () => {
  it('calls synthetic event onClick()', () => {
    const spy = sinon.spy();
    render((
      <Button
        {...mandatoryProps}
        onClick={spy}
      />
    ));

    userEvent.click(screen.getByText('label'));
    expect(spy.calledOnce).toEqual(true);
  });
});
