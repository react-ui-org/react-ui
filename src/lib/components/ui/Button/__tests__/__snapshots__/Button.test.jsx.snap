// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly color 1`] = `
<button
  className="root rootPriorityFilled rootColorSuccess rootSizeMedium    "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    button
  </span>
</button>
`;

exports[`rendering renders correctly flat 1`] = `
<button
  className="root rootPriorityFlat rootColorPrimary rootSizeMedium    "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    button
  </span>
</button>
`;

exports[`rendering renders correctly label visibility 1`] = `
<button
  className="root rootPriorityFilled rootColorPrimary rootSizeMedium withLabelHidden   "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    button
  </span>
</button>
`;

exports[`rendering renders correctly mandatory props only 1`] = `
<button
  className="root rootPriorityFilled rootColorPrimary rootSizeMedium    "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    button
  </span>
</button>
`;

exports[`rendering renders correctly priority \`link\` 1`] = `
<button
  className="root rootPriorityLink"
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    link
  </span>
</button>
`;

exports[`rendering renders correctly priority \`link\` with all props 1`] = `
<button
  className="root rootPriorityLink"
  disabled={true}
  id="custom-id"
  onClick={null}
  type="button"
>
  <span
    className="beforeLabel"
  >
    <span
      className="icon"
    />
  </span>
  <span
    className="label"
    id="custom-id__labelText"
  >
    link
  </span>
  <span
    className="afterLabel"
  >
    <span
      className="icon"
    />
  </span>
</button>
`;

exports[`rendering renders correctly priority 1`] = `
<button
  className="root rootPriorityOutline rootColorPrimary rootSizeMedium    "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    button
  </span>
</button>
`;

exports[`rendering renders correctly size 1`] = `
<button
  className="root rootPriorityFilled rootColorPrimary rootSizeLarge    "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="label"
  >
    button
  </span>
</button>
`;

exports[`rendering renders correctly with all props 1`] = `
<button
  className="root rootPriorityFlat rootColorSuccess rootSizeLarge withLabelHiddenMobile rootBlock rootGrouped isRootLoading"
  disabled={true}
  id="custom-id"
  onClick={null}
  type="button"
>
  <span
    className="startCorner"
  >
    <WithProviderContextComponent
      label={1}
    />
  </span>
  <span
    className="beforeLabel"
  >
    <span
      className="icon"
    />
  </span>
  <span
    className="label"
    id="custom-id__labelText"
  >
    button
  </span>
  <span
    className="afterLabel"
  >
    <span
      className="icon"
    />
  </span>
  <span
    className="endCorner"
  >
    <WithProviderContextComponent
      label={2}
    />
  </span>
  <span
    className="loadingIcon"
  >
    <span
      className="icon"
    />
  </span>
</button>
`;

exports[`rendering renders correctly with icon 1`] = `
<button
  className="root rootPriorityFilled rootColorPrimary rootSizeMedium    "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="beforeLabel"
  >
    <span
      className="icon"
    />
  </span>
  <span
    className="label"
  >
    button
  </span>
</button>
`;
